internal/channel/plugin.go

```diff
@@ -27,60 +110,98 @@ func (c *Channel) GetInfo() (*plugin.Info, error) {
 
 	return err
 }
 
-func (c *Channel) SendNotification(contact *recipient.Contact, i contracts.Incident, ev *event.Event, icingaweb2Url string) error {
-	contactStruct := &plugin.Contact{FullName: contact.FullName}
-	for _, addr := range contact.Addresses {
-		contactStruct.Addresses = append(contactStruct.Addresses, &plugin.Address{Type: addr.Type, Address: addr.Address})
+// SendNotification sends the notification, returns an error if fails
+func (p *Plugin) SendNotification(req *plugin.NotificationRequest) error {
+	params, err := json.Marshal(req)
+	if err != nil {
+		return fmt.Errorf("failed to prepare request params: %w", err)
 	}
 
-	if !strings.HasSuffix(icingaweb2Url, "/") {
-		icingaweb2Url += "/"
+	_, err = p.rpc.Call(plugin.MethodSendNotification, params)
+
+	return err
+}
+
+func forwardLogs(errPipe io.Reader, logger *zap.SugaredLogger) {
+	scanner := bufio.NewScanner(errPipe)
+	for scanner.Scan() {
+		logger.Info(scanner.Text())
 	}
 
-	req := plugin.NotificationRequest{
-		Contact: contactStruct,
-		Object: &plugin.Object{
-			Name:      i.ObjectDisplayName(),
-			Url:       ev.URL,
-			Tags:      ev.Tags,
-			ExtraTags: ev.ExtraTags,
-		},
-		Incident: &plugin.Incident{
-			Id:  i.ID(),
-			Url: fmt.Sprintf("%snotifications/incident?id=%d", icingaweb2Url, i.ID()),
-		},
-		Event: &plugin.Event{
-			Time:     ev.Time,
-			Type:     ev.Type,
-			Severity: ev.Severity.String(),
-			Username: ev.Username,
-			Message:  ev.Message,
-		},
+	if err := scanner.Err(); err != nil {
+		logger.Errorw("Failed to scan stderr line", zap.Error(err))
 	}
+}
 
-	params, err := json.Marshal(req)
+// UpsertPlugins upsert the available_channel_type table with working plugins
+func UpsertPlugins(channelPluginDir string, logger *logging.Logger, db *icingadb.DB) {
+	logger.Debug("Updating available channel types")
+	files, err := os.ReadDir(channelPluginDir)
 	if err != nil {
-		return fmt.Errorf("failed to prepare request params: %w", err)
+		logger.Errorw("Failed to read the channel plugin directory", zap.Error(err))
 	}
 
-	_, err = c.rpcCall(plugin.MethodSendNotification, params)
+	var pluginInfos []*plugin.Info
+	var pluginTypes []string
 
-	return err
+	for _, file := range files {
+		pluginType := file.Name()
+		pluginLogger := logger.With(zap.String("type", pluginType))
+		if err := ValidateType(pluginType); err != nil {
+			pluginLogger.Warnw("Ignoring plugin", zap.Error(err))
+			continue
+		}
+
+		p, err := NewPlugin(pluginType, pluginLogger)
+		if err != nil {
+			pluginLogger.Errorw("Failed to start plugin", zap.Error(err))
+			continue
+		}
+
+		info, err := p.GetInfo()
+		if err != nil {
+			p.logger.Error(err)
+			p.Stop()
+			continue
+		}
+		p.Stop()
+
+		pluginTypes = append(pluginTypes, pluginType)
+		pluginInfos = append(pluginInfos, info)
+	}
+
+	if len(pluginInfos) == 0 {
+		logger.Info("No working plugin found")
+		return
+	}
+
+	stmt, _ := db.BuildUpsertStmt(&plugin.Info{})
+	_, err = db.NamedExec(stmt, pluginInfos)
```

Comment:
Currently, the `type` column is set from the JSON reported by the plugin, but that should always be set to the file name and not even be part of the JSON. (Simply renaming the plugin executable should not introduce any inconsistencies.)